src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -22,6 +23,459 @@ namespace System
+            int length = ClampEnd(span, start, trimElement);
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static Span<T> TrimStart<T>(this Span<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static Span<T> TrimEnd<T>(this Span<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int length = ClampEnd(span, 0, trimElement);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> Trim<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            int length = ClampEnd(span, start, trimElement);
+            return span.Slice(start, length);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimStart<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int start = ClampStart(span, trimElement);
+            return span.Slice(start);
+        }
+
+        /// <summary>
+        /// Removes all trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        public static ReadOnlySpan<T> TrimEnd<T>(this ReadOnlySpan<T> span, T trimElement)
+            where T : IEquatable<T>
+        {
+            int length = ClampEnd(span, 0, trimElement);
+            return span.Slice(0, length);
+        }
+
+        /// <summary>
+        /// Delimits all leading and trailing occurrences of a specified element.
+        /// </summary>
+        /// <param name="span">The source span from which the element is removed.</param>
+        /// <param name="trimElement">The specified element to look for and remove.</param>
+        private static int ClampStart<T>(in ReadOnlySpan<T> span, in T trimElement)
+            where T : IEquatable<T>
+        {
+            if (span.IsEmpty)
+                return 0;
```

Comment:
Do we need the IsEmpty checks? If it's important to optimize for this case, seems like it should be at the call site instead, so that we don't do it twice and make this call, and if it's not important, seems like it could just be removed?

Reply:
I have removed the check altogether - it is not present on the existing `<char>` overloads so I am guessing it was determined that passing an empty span is a low frequency event (which makes sense)